File: examples/python/signal_types.py
"""
CLASP Signal Types Example (Python)

Demonstrates all five signal types:
- Param: Stateful values
- Event: One-shot triggers
- Stream: High-frequency data
- Gesture: Phased input
- Timeline: Automation

Usage:
    python signal_types.py
"""

import asyncio
import os
import math
from clasp_to import Clasp

CLASP_URL = os.environ.get("CLASP_URL", "ws://localhost:7330")


async def main():
    print("=== CLASP Signal Types Example (Python) ===\n")

    client = Clasp(CLASP_URL)
    await client.connect(name="Signal Types Demo")
    print("Connected to CLASP server")

    # =====================
    # 1. PARAM
    # =====================
    print("\n--- 1. PARAM (Stateful Values) ---")

    @client.on("/mixer/**")
    def on_mixer(value, address, meta):
        print(f"[PARAM] {address} = {value} (rev: {meta.get('revision', '?')})")

    await asyncio.sleep(0.1)

    await client.set("/mixer/master/volume", 0.8)
    await client.set("/mixer/master/mute", False)
    await client.set("/mixer/channel/1/volume", 0.65)

    # Get cached value
    master_vol = client.value("/mixer/master/volume")
    print(f"\nCached master volume: {master_vol}")

    await asyncio.sleep(0.3)

    # =====================
    # 2. EVENT
    # =====================
    print("\n--- 2. EVENT (One-Shot Triggers) ---")

    @client.on_event("/cue/**")
    def on_cue(payload, address):
        print(f"[EVENT] {address}: {payload}")

    await asyncio.sleep(0.1)

    await client.emit("/cue/go", {"cue_id": "intro", "fade_time": 2.0})
    await client.emit("/cue/stop", {"immediate": False})

    await asyncio.sleep(0.3)

    # =====================
    # 3. STREAM
    # =====================
    print("\n--- 3. STREAM (High-Frequency Data) ---")

    stream_count = [0]

    @client.on_stream("/sensor/**", max_rate=30)
    def on_sensor(value, address):
        stream_count[0] += 1
        if stream_count[0] % 10 == 0:
            print(f"[STREAM] {address} = {value} (received {stream_count[0]})")

    await asyncio.sleep(0.1)

    # Stream at 60Hz for 1 second
    print("Streaming 60 values...")
    for i in range(60):
        t = i / 60
        await client.stream(
            "/sensor/accelerometer",
            {"x": math.sin(t * math.pi * 2), "y": math.cos(t * math.pi * 2), "z": 0.98},
        )
        await asyncio.sleep(0.016)

    print(f"Sent 60 stream values, received {stream_count[0]}")

    await asyncio.sleep(0.3)

    # =====================
    # 4. GESTURE
    # =====================
    print("\n--- 4. GESTURE (Phased Input) ---")

    @client.on_gesture("/input/**")
    def on_gesture(gesture):
        phase = gesture.get("phase", "?")
        x = gesture.get("x", 0)
        y = gesture.get("y", 0)
        print(f"[GESTURE] {phase.upper()} at ({x:.0f}, {y:.0f})")

    await asyncio.sleep(0.1)

    gesture_id = f"drag-{int(asyncio.get_event_loop().time() * 1000)}"

    await client.gesture("/input/mouse", {"id": gesture_id, "phase": "start", "x": 100, "y": 100})

    for i in range(1, 6):
        await client.gesture(
            "/input/mouse", {"id": gesture_id, "phase": "move", "x": 100 + i * 20, "y": 100 + i * 10}
        )
        await asyncio.sleep(0.05)

    await client.gesture("/input/mouse", {"id": gesture_id, "phase": "end", "x": 200, "y": 150})

    await asyncio.sleep(0.3)

    # =====================
    # 5. TIMELINE
    # =====================
    print("\n--- 5. TIMELINE (Automation) ---")

    @client.on_timeline("/automation/**")
    def on_timeline(timeline, address):
        kf_count = len(timeline.get("keyframes", []))
        duration = timeline.get("duration", 0)
        print(f"[TIMELINE] {address}: {kf_count} keyframes, duration={duration}ms")

    await asyncio.sleep(0.1)

    await client.timeline(
        "/automation/light/brightness",
        {
            "duration": 5000,
            "loop": True,
            "keyframes": [
                {"time": 0, "value": 0.0, "easing": "linear"},
                {"time": 1000, "value": 1.0, "easing": "ease-out"},
                {"time": 3000, "value": 1.0, "easing": "linear"},
                {"time": 5000, "value": 0.0, "easing": "ease-in"},
            ],
        },
    )

    await asyncio.sleep(0.3)

    # Summary
    print("\n=== Signal Type Summary ===")
    print("| Type     | QoS     | Persists | Use Case                    |")
    print("|----------|---------|----------|-----------------------------|")
    print("| Param    | Confirm | Yes      | Faders, settings, state     |")
    print("| Event    | Confirm | No       | Button press, cue trigger   |")
    print("| Stream   | Fire    | No       | Sensors, meters (30-60Hz)   |")
    print("| Gesture  | Fire    | No       | Touch, pen, mouse drag      |")
    print("| Timeline | Commit  | Yes      | Animation, automation       |")

    await client.disconnect()
    print("\n=== Signal types demo complete ===")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/js/signal-types.js
/**
 * CLASP Signal Types Example
 *
 * Demonstrates all five signal types in CLASP:
 * - Param: Stateful values that persist and sync
 * - Event: One-shot triggers that don't persist
 * - Stream: High-frequency data (lossy, fast)
 * - Gesture: Phased input (start/move/end)
 * - Timeline: Automation lanes with keyframes
 *
 * Usage:
 *   node signal-types.js
 */

import { Clasp } from '@clasp-to/core';

const CLASP_URL = process.env.CLASP_URL || 'ws://localhost:7330';

async function main() {
  console.log('=== CLASP Signal Types Example ===\n');

  const client = new Clasp(CLASP_URL);
  await client.connect({ name: 'Signal Types Demo' });
  console.log('Connected to CLASP server');

  // =====================
  // 1. PARAM - Stateful Values
  // =====================
  console.log('\n--- 1. PARAM (Stateful Values) ---');
  console.log('Params persist on the server and sync to late joiners.\n');

  // Subscribe to params
  client.on('/mixer/**', (value, address, meta) => {
    console.log(`[PARAM] ${address} = ${JSON.stringify(value)} (rev: ${meta.revision})`);
  });

  await new Promise(r => setTimeout(r, 100));

  // Set some param values
  await client.set('/mixer/master/volume', 0.8);
  await client.set('/mixer/master/mute', false);
  await client.set('/mixer/channel/1/volume', 0.65);
  await client.set('/mixer/channel/1/pan', 0.0);
  await client.set('/mixer/channel/1/eq', { low: 0, mid: 2, high: -1 });

  // Get current value (sync - uses cached value)
  const masterVol = client.value('/mixer/master/volume');
  console.log(`\nCached master volume: ${masterVol}`);

  // Get current value (async - fetches from server)
  const channel1 = await client.get('/mixer/channel/1/volume');
  console.log(`Fetched channel 1 volume: ${channel1.value} (rev: ${channel1.revision})`);

  await new Promise(r => setTimeout(r, 300));

  // =====================
  // 2. EVENT - One-Shot Triggers
  // =====================
  console.log('\n--- 2. EVENT (One-Shot Triggers) ---');
  console.log('Events fire once and are not stored. Miss it and it\'s gone.\n');

  // Subscribe to events
  client.onEvent('/cue/**', (payload, address) => {
    console.log(`[EVENT] ${address}:`, payload);
  });

  client.onEvent('/button/**', (payload, address) => {
    console.log(`[EVENT] ${address} pressed`);
  });

  await new Promise(r => setTimeout(r, 100));

  // Emit events
  await client.emit('/cue/go', { cueId: 'intro', fadeTime: 2.0 });
  await client.emit('/cue/stop', { immediate: false });
  await client.emit('/button/play', null);
  await client.emit('/button/record', { armed: true });

  // Events with metadata
  await client.emit('/notification/alert', {
    title: 'Warning',
    message: 'CPU usage high',
    level: 'warning'
  });

  await new Promise(r => setTimeout(r, 300));

  // =====================
  // 3. STREAM - High-Frequency Data
  // =====================
  console.log('\n--- 3. STREAM (High-Frequency Data) ---');
  console.log('Streams are lossy but fast. Use for sensors, meters, etc.\n');

  let streamCount = 0;

  // Subscribe to streams with rate limiting
  client.onStream('/sensor/**', (value, address) => {
    streamCount++;
    if (streamCount % 10 === 0) {
      console.log(`[STREAM] ${address} = ${JSON.stringify(value)} (received ${streamCount} total)`);
    }
  }, { maxRate: 30 }); // Limit to 30 updates/sec

  await new Promise(r => setTimeout(r, 100));

  // Stream sensor data at 60Hz for 1 second
  console.log('Streaming 60 values...');
  for (let i = 0; i < 60; i++) {
    const t = i / 60;
    await client.stream('/sensor/accelerometer', {
      x: Math.sin(t * Math.PI * 2),
      y: Math.cos(t * Math.PI * 2),
      z: 0.98 + Math.random() * 0.02
    });

    await client.stream('/sensor/temperature', 22.5 + Math.random() * 0.5);

    await new Promise(r => setTimeout(r, 16));
  }

  console.log(`Sent 120 stream values, received ${streamCount} (rate limited)`);

  await new Promise(r => setTimeout(r, 300));

  // =====================
  // 4. GESTURE - Phased Input
  // =====================
  console.log('\n--- 4. GESTURE (Phased Input) ---');
  console.log('Gestures have start/move/end phases. Used for touch, pen, drag.\n');

  // Subscribe to gestures
  client.onGesture('/input/**', (gesture) => {
    const { id, phase, x, y } = gesture;
    console.log(`[GESTURE] ${phase.toUpperCase()} id=${id.slice(-6)} at (${x?.toFixed(0)}, ${y?.toFixed(0)})`);
  });

  await new Promise(r => setTimeout(r, 100));

  // Simulate a drag gesture
  const gestureId = `drag-${Date.now()}`;

  await client.gesture('/input/mouse', { id: gestureId, phase: 'start', x: 100, y: 100 });

  for (let i = 1; i <= 5; i++) {
    await client.gesture('/input/mouse', {
      id: gestureId,
      phase: 'move',
      x: 100 + i * 20,
      y: 100 + i * 10
    });
    await new Promise(r => setTimeout(r, 50));
  }

  await client.gesture('/input/mouse', { id: gestureId, phase: 'end', x: 200, y: 150 });

  await new Promise(r => setTimeout(r, 300));

  // =====================
  // 5. TIMELINE - Automation
  // =====================
  console.log('\n--- 5. TIMELINE (Automation Lanes) ---');
  console.log('Timelines store keyframes for automated playback.\n');

  // Subscribe to timeline updates
  client.onTimeline('/automation/**', (timeline, address) => {
    console.log(`[TIMELINE] ${address}: ${timeline.keyframes.length} keyframes, duration=${timeline.duration}ms`);
  });

  await new Promise(r => setTimeout(r, 100));

  // Create a timeline with keyframes
  await client.timeline('/automation/light/brightness', {
    duration: 5000, // 5 seconds
    loop: true,
    keyframes: [
      { time: 0, value: 0.0, easing: 'linear' },
      { time: 1000, value: 1.0, easing: 'ease-out' },
      { time: 3000, value: 1.0, easing: 'linear' },
      { time: 4000, value: 0.3, easing: 'ease-in-out' },
      { time: 5000, value: 0.0, easing: 'linear' }
    ]
  });

  // Create a color timeline
  await client.timeline('/automation/light/color', {
    duration: 5000,
    loop: true,
    keyframes: [
      { time: 0, value: { r: 255, g: 0, b: 0 }, easing: 'linear' },
      { time: 2500, value: { r: 0, g: 255, b: 0 }, easing: 'linear' },
      { time: 5000, value: { r: 255, g: 0, b: 0 }, easing: 'linear' }
    ]
  });

  // Control timeline playback
  await client.emit('/automation/light/play', { startTime: client.time() });

  await new Promise(r => setTimeout(r, 500));

  // Query timeline state
  const brightnessTL = await client.getTimeline('/automation/light/brightness');
  console.log('\nBrightness timeline:', brightnessTL);

  await new Promise(r => setTimeout(r, 300));

  // =====================
  // Summary
  // =====================
  console.log('\n=== Signal Type Summary ===');
  console.log('| Type     | QoS     | Persists | Use Case                    |');
  console.log('|----------|---------|----------|-----------------------------|');
  console.log('| Param    | Confirm | Yes      | Faders, settings, state     |');
  console.log('| Event    | Confirm | No       | Button press, cue trigger   |');
  console.log('| Stream   | Fire    | No       | Sensors, meters (30-60Hz)   |');
  console.log('| Gesture  | Fire    | No       | Touch, pen, mouse drag      |');
  console.log('| Timeline | Commit  | Yes      | Animation, automation       |');

  await client.disconnect();
  console.log('\n=== Signal types demo complete ===');
}

main().catch(console.error);
